[GTK] gsk_gl_command_queue_create_render_target: assertion failed (glCheckFramebufferStatus (GL_FRAMEBUFFER) == GL_FRAMEBUFFER_COMPLETE) when opening webkit://gpu

carlosgcampos opened the pull request:
#### edc3fa95e5155cd31a53ef133d6018d052754c39
<pre>
[GTK] gsk_gl_command_queue_create_render_target: assertion failed (glCheckFramebufferStatus (GL_FRAMEBUFFER) == GL_FRAMEBUFFER_COMPLETE) when opening webkit://gpu
<a href="https://bugs.webkit.org/show_bug.cgi?id=256873">https://bugs.webkit.org/show_bug.cgi?id=256873</a>

Reviewed by Michael Catanzaro.

Add ScopedGLContext class to create a GLContext that is made current for
the current scope and restores the previous current context on destruction.
In case of GLX being used by the UI process, use a dedicated thread to
create the EGL context and fill the GPU information.

* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::ScopedGLContext::ScopedGLContext):
(WebCore::GLContext::ScopedGLContext::~ScopedGLContext):
* Source/WebCore/platform/graphics/egl/GLContext.h:
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::WebKitProtocolHandler::handleGPU):

Canonical link: <a href="https://commits.webkit.org/264311@main">https://commits.webkit.org/264311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/263c4a0086d0b92155876c522c1812913e6b9a34

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7159 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7407 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8776 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7379 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7169 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7340 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10282 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7286 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7963 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6595 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8882 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5314 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6514 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14253 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6965 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9490 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7095 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5797 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6455 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10654 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/854 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6838 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

benjamin-otte commented on Source/WebCore/platform/graphics/egl/GLContext.cpp:
```diff
@@ -640,6 +640,25 @@ unsigned GLContext::version()
+        m_previous.readSurface = eglGetCurrentSurface(EGL_READ);
+        m_previous.drawSurface = eglGetCurrentSurface(EGL_DRAW);
+    }
+    m_context->makeContextCurrent();
```
The only reason where I can imagine this failing with GTK is if GTK doesn't use EGL but GLX.

carlosgcampos replied:
If GTK is using GLX there's no current EGLContext, I guess

benjamin-otte replied:
Right, in that case there'd be a current GLX context.

Calling `gdk_gl_context_clear_current()` would get rid of either, and GTK will call `make_current()` again, so calling that  is fine.
The other thing is that Webkit would need to make sure to call `clear_current()` (aka `eglMakeCurrent(NULL)`) once it's done so that GTK (if using GLX) doesn't cause problems when it calls make_current() later.

That reminds me: You might be able to reproduce this on X11 with `GDK_DEBUG=gl-glx` or `GDK_DEBUG=gl-egl` depending on what the problem here is - if it's the EGL/GLX interaction it should work, but if it requires a specific driver to cause that to fall over, then that won't be enough to trigger the fault.

carlosgcampos replied:
I'm not sure I understand... Does eglMakeCurrent affect a possible GLX current context? Note that we don't support GLX anymore in WebKit, so when GTK uses GLX (which is always for me under X11 because mesa doesn't provide any argb32 egl config) we fallback to cairo for rendering.

benjamin-otte replied:
Okay, lemme explain the madness that is GL as far as I understand it:

When you call `glFoo()`, the library needs to decide what to call. Keep in mind that you can call `glFoo()` both with EGL and with GLX (and with any other GL platform library), so behind that function there is (depending on the libGL you are using, ie Mesa or nvidia or Microsoft or...) some dispatch table that figures out the right function to call.

However, it is undefined how that dispatch table works. It is assumed that when you call `make_current()`, the dispatch table gets set up for the context you pass in there.  
But if you mix and match calls to `eglMakeCurrent()` and `glXMakeCurrent()`, each of those may (or may not) turn into no-ops if the context you are setting is already the current egl context or current glx context, which means they may or may not replace the dispatch table in that case.

Consider this code:
```c
glXMakeCurrent(gdk->glx_context); /* called by the GTK renderer */
eglMakeCurrent(webkit->egl_context); /* called by your Webkit code */
glXMakeCurrent(gdk->glx_context);  /* called by the GTK renderer next frame */
```
Does that last call now reset the global dispatch table that makes sure that `glFoo()` points to the right function?  

And I think different drivers do different things here, so reading the Mesa impl might not be enough to understand what's going on.

carlosgcampos replied:
Thanks for the explanation, it's clearer now. So, all this current context is per thread, right? I think that another possible solution might be to create and destroy the temporal EGL offscreen contexts in a separate thread. This is only used for webkit://gpu so, it should be ok to block the main thread while waiting for the EGL context information to be filled.

benjamin-otte replied:
Yes, this is per thread.

If one thread uses GLX and the other uses EGL, things should be working fine - and probably have been with webkitgtk on X11 for a while.

carlosgcampos committed:
[GTK] gsk_gl_command_queue_create_render_target: assertion failed (glCheckFramebufferStatus (GL_FRAMEBUFFER) == GL_FRAMEBUFFER_COMPLETE) when opening webkit://gpu
https://bugs.webkit.org/show_bug.cgi?id=256873

Reviewed by Michael Catanzaro.

Add ScopedGLContext class to create a GLContext that is made current for
the current scope and restores the previous current context on destruction.
In case of GLX being used by the UI process, use a dedicated thread to
create the EGL context and fill the GPU information.

* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::ScopedGLContext::ScopedGLContext):
(WebCore::GLContext::ScopedGLContext::~ScopedGLContext):
* Source/WebCore/platform/graphics/egl/GLContext.h:
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::WebKitProtocolHandler::handleGPU):

Canonical link: https://commits.webkit.org/264311@main
